src/analyses/base.ml

```diff
@@ -934,6 +934,52 @@ struct
         (*          Messages.warn ~msg:("Base: I should know it! "^string_of_int (List.length fs)) ();*)
         List.fold_left (fun xs v -> Q.LS.add (v,`NoOffset) xs) (Q.LS.empty ()) fs
       end
+    | Q.Assert e ->
+      begin
+        let i = query_evalint (Analyses.ask_of_ctx ctx) ctx.global ctx.local e in
+        let tmp = match e with | Lval l -> Some l | _ -> None in
+        let context: Invariant.context = {
+          scope=MyCFG.dummy_func;
+          i= -1; (* Not used here *)
+          lval=tmp;
+          offset=Cil.NoOffset;
+          deref_invariant=(fun _ _ _ -> Invariant.none) (* TODO: should throw instead? *)
+        }
+        in
+        match (VD.invariant context (`Int i)) with
+        | Some s -> Queries.ES.singleton s
+        | None -> `Top
+      end
```

Comment:
I'm quite confused what this query is trying to achieve and what it's even for.

It first evaluates the expression to an `Int` abstraction. Then it uses `VD.invariant` to convert that back to an expression and return it. Why all this trouble instead of just calling `D.invariant`, which constructs an invariant expression for the entire state.

And where is this expression even coming from? Is it just a single variable name? It cannot be anything else because `VD.invariant` will be an expression only about a single lvalue.

Also, why is it just in base analysis? The other analyses may have invariants too. Really `Spec.D.invariant` should be called on the top level of the analysis and that already takes care of everything. That's how witness generation gets invariants.

Reply:
> Why all this trouble instead of just calling D.invariant, which constructs an invariant expression for the entire state.

The idea is that one does not want to assert everything about the local state everywhere, and instead assert only the values of some interesting lvals, e.g. the one that is assigned to, in order to not pollute the program with too many asserts.

> Also, why is it just in base analysis? The other analyses may have invariants too

I guess all invariants about this single variable would be obtained via this `query_evalint` here, as that in turn queries all other analyses, right?

Reply:
> The idea is that one does not want to assert everything about the local state everywhere, and instead assert only the values of some interesting lvals, e.g. the one that is assigned to, in order to not pollute the program with too many asserts.

Then at minimum the query argument should be `lval` instead of `exp`.

Also, the context for `D.invariant` has a field for choosing which `lval` to output it for, so that can already be set to get the invariant for a particular variable without it having to go through a query, which just calls `invariant` exactly that way.


> I guess all invariants about this single variable would be obtained via this `query_evalint` here, as that in turn queries all other analyses, right?

Not necessarily. That would only give invariants representable by `ID`. I think Apron analysis doesn't have `invariant` implemented right now, but it could also extract its constraints, so one might want to output an invariant `x < y` for `lval = x`.

Reply:
> without it having to go through a query, which just calls invariant exactly that way

The point is that these internals are not exposed to the transformation system, the only thing that one has available there is the query system (and that is probably a good design). 

I modified it such that we now have a `Invariant` query that we call directly from the transformation and answer in `MCP2`, but it always returns `Invariant.none` so far :thinking:

Reply:
Ok, the key was providing the proper scope.

Reply:
> Also, the context for D.invariant has a field for choosing which lval to output it for, so that can already be set to get the invariant for a particular variable without it having to go through a query, which just calls invariant exactly that way.

Is that what the intended use of that `lval option` is? Because thus far, it is simply overwritten e.g. in `BaseDomain.invariant` such that the invariant is produced for all bindings in the map.

Reply:
Making an `Invariant` query is a good idea. I think also thought of doing that at some point (and possibly removing `invariant` from `Printable.S`, but I don't remember anymore why I didn't do it. Probably because there's some involved recursive logic for the invariants of address and struct domains. I might have to look into using this query for witnesses though.

The original use of `lval` in the invariant context was to allow, e.g. interval domain to know which variable to use for the interval invariant expression it creates. But I think then somewhere it also got used for picking particular things, although maybe not on the `CPA` level. The struct domain seems to do something similar (although for `offset`), but this should be fine to reuse for the different purpose outside of `CPA`.